### ISSUE TITLE
Allow customizations to the User Profile page

### ISSUE THREAD

**Issue by heatheralee** [backstage/backstage#5382]
## Feature Suggestion
Hi there! I'm currently working on a feature which will allow users to update their user profile with more details that they can fill out in the User Settings page. I wanted to see if this is something I could integrate and contribute to the Backstage codebase/what sorts of things I could make configurable so that this is helpful to everyone. An example flow would look like:

1. I update my profile on my user settings page: heatheralee. I add information like: my preferred modes of communication, my time zone, a short blurb about me.
2. Someone else searches for the UserEntity heatheralee and finds my UserProfileCard as well as the other information I've added in step 1.

## Possible Implementation

1. Allow users to decide which fields users can author about themselves by adding field name and field type to app-config.yaml file. These fields will come from a predefined set (tbd).
2. Update `plugins/user-settings/src/components/SettingsPage.tsx` to iterate through app configuration in step 1 and render form fields.
3. Once user fills out these fields, update the UserEntity `profile` to contain added information.
4. Update `plugins/org/src/plugin.ts` to include new cards which render custom fields, and allow users to add this in `packages/app/src/components/catalog/EntityPage.jsx`.

## Context

With this change, we are enhancing the user profiles with more information. One example of why this is helpful so that when users find owners of services, and they click on the owner username, they can get more information about the owner.

**Comment by younky-yang** [comment]
Like the idea, but first we want to enable the user-settings page to populate more information that we can query from the auth provider. This may related to the identityApi which currently is very basic info included.

**Comment by evansche** [comment]
I especially like the DevEx benefits entailed in _"when users find owners of services, and they click on the owner username, they can get more information about the owner"_. That's a powerful community multiplier. Sourcing from both the auth provider and users' own supplemental enrichment sounds ideal.

**Comment by heatheralee** [comment]
Thanks for both your comments!

> Like the idea, but first we want to enable the user-settings page to populate more information that we can query from the auth provider. This may related to the identityApi which currently is very basic info included.

Yes agreed -- the more we can pull from the auth provider the better! 
I would like a way to enhance the existing data we get from the auth provider like @evansche mentioned, as depending on the auth provider, you may not get that much information back (we are currently using LDAP). For those using an auth provider with more information included, it would be nice to still offer the functionality to leverage both the default information provided and also allow users to author content about themselves to fill in any gaps where data could not be pulled.

**Comment by adamdmharvey** [comment]
Could be value integrating the existing GitHub User README feature here as well:
https://docs.github.com/en/github/setting-up-and-managing-your-github-profile/managing-your-profile-readme

**Comment by tonytamsf** [comment]
> I especially like the DevEx benefits entailed in _"when users find owners of services, and they click on the owner username, they can get more information about the owner"_. That's a powerful community multiplier. Sourcing from both the auth provider and users' own supplemental enrichment sounds ideal.

We are seeing more internal developers creating a README about themselves and how best to collaborate with them.  Having these README linked on their backstage.io profile is powerful.  On anther issue, we will propose having a Bravo badge system, were people can use backstage.io give badges of appreciation.

**Comment by timbonicus** [comment]
I agree wholeheartedly with this idea! I think the suggested implementation may be troublesome since Backstage does not want to be the source of truth for user information - that should instead be your user management system (LDAP, Github teams, etc) that is then wired into Backstage. I'd say it's best for User entities to be loaded from that system with as little manipulation as possible, so that the source of truth is maintained.

That being said, it makes total sense for user management to be done through the Backstage UI (since this is just another infrastructure tool, which Backstage shines in bringing together). I imagine this could be a high-level User or Group editor that supports different integrations for common user management backends. Then, as others mentioned here, your user profile in that backend system is transformed for frontend display. Maybe this is your LDAP profile, Github README, etc. This could even be edited in Backstage but have data posted to the backend system. We have exactly this internally at Spotify.

There's some distinction to be made between the `SettingsPage` mentioned, which shows information for _my own logged in identity_, and [User entity](https://demo.backstage.io/catalog/default/user/breanna.davison) pages from the catalog, which can show user information for **any** user in the catalog.

> Like the idea, but first we want to enable the user-settings page to populate more information that we can query from the auth provider. This may related to the identityApi which currently is very basic info included.

I think this may be misguided - auth providers are meant to handle logging in users, not populating Users in the catalog. There's conceptually confusing overlap since some auth providers also have extended profile information for users and groups, but Backstage is designed to ingest User catalog entities with [processors](https://backstage.io/docs/features/software-catalog/configuration#processors) rather than auth providers. Using the latter would imply that you only have full user profile information for users that have logged into the system.

**Comment by heatheralee** [comment]
Thanks @adamdmharvey and @timbonicus for your feedback!

Yes, after poking around some more in the code yesterday, I realized that updating the User Entity after the data had been ingested in `refreshAllLocations()` didn't feel very natural to the system. I found a [POST /entities endpoint](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/service/router.ts#L93) in plugins/catalog-backend but noticed in the comments that it was deprecated and not recommended for use. 

> That being said, it makes total sense for user management to be done through the Backstage UI (since this is just another infrastructure tool, which Backstage shines in bringing together). I imagine this could be a high-level User or Group editor that supports different integrations for common user management backends. Then, as others mentioned here, your user profile in that backend system is transformed for frontend display. Maybe this is your LDAP profile, Github README, etc. This could even be edited in Backstage but have data posted to the backend system. We have exactly this internally at Spotify.

I love this and would love to hear more about the internal implementation at Spotify (if possible)! I imagine this backend system lives outside the Backstage ecosystem? Does the high-level User or Group editor live outside Backstage as well?

**Comment by timbonicus** [comment]
> I love this and would love to hear more about the internal implementation at Spotify (if possible)! I imagine this backend system lives outside the Backstage ecosystem? Does the high-level User or Group editor live outside Backstage as well?

Yes, we have a separate system that pulls together and synchronizes LDAP, Google users/groups and GCP IAM. This system exposes an API, then the management UI is a Backstage frontend plugin that calls that API to make changes.

**Comment by heatheralee** [comment]
> We have a separate backend system that pulls together and synchronizes LDAP, Google users/groups and GCP IAM. This system exposes an API, then the high-level User/Group management UI is entirely within Backstage as a frontend plugin that calls that API to make changes.

That makes sense, thank you. Does the high-level User/Group management UI act as a replacement for `plugins/org`? Or was `plugins/org` updated to render the newly enriched data from this API? I noticed that UserProfileCard, for example, shows Email, Member Of, and Ownership cards, but would like to get to a state where this card renders whatever is received from this external backend system. I'm wondering if I should dedicate a new plugin for this or update the existing `plugins/org` code.

**Comment by timbonicus** [comment]
Spotify has its own internal plugin for this; not for any meaningful reason, simply because our internal Backstage existed well before the open-source version and we're still in a slow process of applying certain concepts internally.

I think this could definitely make sense as an extension of the `org` plugin, though would require an RFC with some careful deliberation to make sure a proposed architecture is suitably flexible, adaptable and degrades nicely when a backend system isn't in place.

**Comment by heatheralee** [comment]
> I think this could definitely make sense as an extension of the org plugin, though would require an RFC with some careful deliberation to make sure a proposed architecture is suitably flexible, adaptable and degrades nicely when a backend system isn't in place.

Great, thank you. I will take a stab at creating an RFC and follow up in this thread.

**Comment by heatheralee** [comment]
Hi @timbonicus! I'm currently exploring our options/doing an investigation on how best to approach this, and would love your input on if this backend system should:

1. Host the code to ingest data from various different sources (even though Backstage already has this functionality and ingests users and groups from LDAP, for example). I'm hesitant to move forward with this approach as it feels like re-inventing the wheel, but it also feels cleaner than the next option.
2. Leverage Backstage's existing code to pull data from various sources, then either 
     a. send this data from Backstage to the backend system via API calls so that the backend system can merge all of these data sources together, or 
     b. enhance the existing data pulled in Backstage by making API requests to this new service in some place like `preProcessEntity`

Thanks in advance!

**Comment by timbonicus** [comment]
Hmm, let me clarify my thoughts and come back to your question.

I imagined extending `plugin-org` to add a group/user management _frontend_ in Backstage that's adaptable to different backend systems. The backend system may be LDAP, GitHub, Okta, or a custom API such as at Spotify.

The concept of [integrations](https://backstage.io/docs/integrations/) enable Backstage to read **or publish** data to various providers. Reading org data is already provided in some cases, such as the `GithubOrgReaderProcessor`. The publish part of the concept is only used by software templates so far, but I can imagine integrations optionally providing org **write** capability.

So imagine the UI emits a `User` or `Group` entity, with appropriate relations, which is passed to a write-supporting org integration to interpret and publish the changes to the external system. The updated entity is then read back through the normal catalog process.

This wouldn't be ideal at the moment, since the catalog refresh runs at a fixed rate and wouldn't pick up the external change right away. That will soon change with the [catalog/next](https://github.com/backstage/backstage/tree/master/plugins/catalog-backend/src/next) work. It has a concept of `EntityProvider` which allows entity updates more realtime than they are today.

So to get back to your question - in this scenario, Backstage still pulls data from org integrations as in (2), and supports edit/push via write-friendly org integrations which asynchronously update the catalog once complete. There's a slight, hopefully minimal, delay for changes to be reconciled. Or perhaps the write integration could interact with the catalog to trigger an immediate, synchronous refresh for the entity.

I paint this picture imagining a single org integration and data source. I think you're alluding to a situation with multiple org data sources. In Spotify's case, even though there are multiple _underlying_ data sources, this is still managed through a single backend system that synchronizes them all and provides a write API, so there's no further complication.

A situation with actual multiple org data source becomes less clear. Are `User` and `Group` entities ingested with an annotation that indicates their backing source, so updates can be pushed to the appropriate provider? Something like [edit-url](https://backstage.io/docs/features/software-catalog/well-known-annotations#backstageioview-url-backstageioedit-url)? We would only allow a `User` ingested from one provider to be added to a `Group` from the same provider? There are some questions to be answered.

**Comment by heatheralee** [comment]
Hi @timbonicus! I wrote an RFC to discuss this in a bit more detail: https://github.com/backstage/backstage/issues/5930. Would love to get your thoughts when you get the chance!
